contracts/sol4/fprConversionRate/ConversionRateEnhancedSteps2.sol

```diff
@@ -0,0 +1,77 @@
+pragma solidity 0.4.18;
+
+import "./ConversionRateEnhancedSteps.sol";
+import "../ERC20Interface.sol";
+
+/* solhint-disable max-line-length */
+contract ConversionRateEnhancedSteps2 is ConversionRateEnhancedSteps {
+    uint256[] internal emptySlotIndicies;
+
+    event AddToken(
+        ERC20 indexed token,
+        uint256 compactDataArrayIndex,
+        uint256 compactDataFieldIndex
+    );
+    event RemoveToken(ERC20 indexed token, uint256 emptyIndexSlot);
+
+    function ConversionRateEnhancedSteps2(address _admin)
+        public
+        ConversionRateEnhancedSteps(_admin)
+    {}
+
+    /// @dev add a token to reserve, if there is empty slots, fill it else create a new slot
+    function addToken(ERC20 token) public onlyAdmin {
+        require(!tokenData[token].listed);
+        tokenData[token].listed = true;
+
+        uint256 compactDataArrayIndex;
+        uint256 compactDataFieldIndex;
+
+        if (emptySlotIndicies.length != 0) {
+            // pop the last empty slot
+            uint256 slotIndex = emptySlotIndicies[emptySlotIndicies.length - 1];
+            emptySlotIndicies.length--;
+
+            compactDataArrayIndex = slotIndex / NUM_TOKENS_IN_COMPACT_DATA;
+            compactDataFieldIndex = slotIndex % NUM_TOKENS_IN_COMPACT_DATA;
+
+            listedTokens[slotIndex] = token;
+        } else {
+            if (numTokensInCurrentCompactData == 0) {
+                tokenRatesCompactData.length++; // add new structure
+            }
+            compactDataArrayIndex = tokenRatesCompactData.length - 1;
+            compactDataFieldIndex = numTokensInCurrentCompactData;
+            // prettier-ignore
+            numTokensInCurrentCompactData = (numTokensInCurrentCompactData + 1) % NUM_TOKENS_IN_COMPACT_DATA;
+
+            listedTokens.push(token);
+        }
+
+        tokenData[token].compactDataArrayIndex = compactDataArrayIndex;
+        tokenData[token].compactDataFieldIndex = compactDataFieldIndex;
+
+        setGarbageToVolumeRecorder(token);
+        setDecimals(token);
+
+        AddToken(token, compactDataArrayIndex, compactDataFieldIndex);
+    }
+
+    /// @dev remove a token from compact data
+    function removeToken(ERC20 token) public onlyAdmin {
+        require(tokenData[token].listed);
+        TokenData storage data = tokenData[token];
+        // prettier-ignore
+        uint256 slotIndex = data.compactDataArrayIndex * NUM_TOKENS_IN_COMPACT_DATA + data.compactDataFieldIndex;
+        emptySlotIndicies.push(slotIndex);
+        // disable token and remove it from listedTokens
+        data.listed = false;
```

Comment:
pls also disable token trade to reduce risk in case we relist that token
`data.enabled = false`

Reply:
change as request